src/drivers/esp8266_drv.c

```diff
@@ -127,102 +190,163 @@ static const char* get_channel_name(const size_t chan_id)
 
-        /* Call back to listening task to indicate new socket */
-        if (!state.new_conn_cb) {
-                pr_error(_LOG_PFX "No Serial callback defined\r\n");
-                ch->in_use = false; /* Channel is no longer in use */
+        /*
+         * Since these connections are created by an external source,
+         * we set the created_externally flag here.  This way
+         * we know to reap the channel when the connection disappears.
+         */
+        ch->created_externally = true;
+        ch->in_use = true;
+        ch->connected = true;
+
+        /*
+         * Check that we actually have a call back set to handle the
+         * incomming data.  If not then the socket stays open until it is
+         * closed by our WiFi host.
+         */
+        if (!state.comm.new_conn_cb) {
+                pr_error(LOG_PFX "No Serial callback defined\r\n");
                 return;
         }
-        state.new_conn_cb(ch->serial);
 
-        pr_info_str_msg(_LOG_PFX "Message: ", data);
-        /* STIEG: Make this not portMAX_DELAY to prevent stalling? */
+        state.comm.new_conn_cb(ch->serial);
+
+        /*
+         * Now that the upper layer has been informed about the incomming
+         * data, start sending it said data.  It will unblock and read this
+         * data very shortly.
+         */
+        /* pr_info_int_msg(LOG_PFX "rx_data_cb data len: ", len); */
+        /* pr_info(LOG_PFX "rx_data_cb data: \""); */
         xQueueHandle q = serial_get_rx_queue(ch->serial);
-        for (size_t i = 0; i < len; ++i)
+        for (size_t i = 0; i < len; ++i) {
+                /* pr_info_int(data[i]); */
+                /* pr_info_char('|'); */
                 xQueueSend(q, data + i, portMAX_DELAY);
+        }
+        /* pr_info("\"\r\n"); */
+
+        cmd_set_check(CHECK_DATA);
+}
+
+/**
+ * Callback that is invoked when the send_data method completes.
+ */
+static void _send_data_cb(int bytes_sent)
+{
+        cmd_completed();
+        cmd_set_check(CHECK_DATA);
+
+        if (bytes_sent < 0)
+                /* STIEG: Include channel info here somehow */
+                pr_warning(LOG_PFX "Failed to send data\r\n");
+}
+
+/**
+ * Method that processes outgoing data if any.  If there is, this
+ * will start a command.
+ */
+static void check_data()
+{
+        cmd_check_complete(CHECK_DATA);
+
+        for (size_t i = 0; i < ARRAY_LEN(state.comm.channels); ++i) {
+                struct channel *ch = state.comm.channels + i;
+                const size_t size = ch->tx_chars_buffered;
+
+                /* If the size is 0, nothing to send */
+                if (0 == size)
+                        continue;
+
+                /*
+                 * If here, then we have data to send.  Check that the
+                 * connection is still active.  If not, dump the data
+                 */
+                if (!ch->connected) {
+                        char c;
```

Comment:
So we emulate a Serial device over a socket in this code. Since serial devices don't check open/closed state, we need to handle the case where a Serial device is still open but has no backing connection.  This is that case.  If the connection is broken for whatever reason and we wait for it to come back (which it won't), we will block indefinitely, causing a permanent stall and eventually a watchdog.  This code says that if we have lost our connection, do exactly what a serial device would do with a broken connection: send data to a non-existent endpoint, which is effectively the same as dropping the data.  This prevents stalls and keeps the overall system in good health.  Once we use proper sockets on top of Serial instead of Serial on top of sockets this can die... but until then this is a necessary evil to keep our changes needed small.